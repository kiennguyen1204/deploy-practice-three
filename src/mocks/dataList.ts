export const dataItemDetail = {
  effects: [
    {
      title: 'effects',
      image: '/src/assets/images/products/color_filter.png',
      content: 'Calming, Creative, Happy, Relaxing, Sleepy, Uplifting'
    },
    {
      title: 'relieve',
      image: '/src/assets/images/products/brain.png',
      content:
        'Anxiety, Arthritis, Chronic Pain, Depression, Fatigue, Inflammation, Insomnia, Irregular Bowel Movements, Migraines, Mood Swings'
    },
    {
      title: 'aromas',
      image: '/src/assets/images/products/carbon_smoke.svg',
      content: 'Chemical, Citrus, Earthy, Pungent, Sour'
    }
  ]
};

export const DATA_LIST = [
  {
    isSoldOut: 86,
    title: '',
    totalReviews: 135,
    discount: '$200.00',
    price: '$102.00',
    unit: 'unit 1',
    volume: [],
    image: 'https://tinyurl.com/2s34y377',
    sale: 'Indica 70%',
    description:
      'Jungle Diamonds is a slightly indica dominant hybrid strain (60% indica/40% sativa) created through crossing the infamous Slurricane X Gorilla Glue #4 strains.',
    category: 'CONCENTRATES',
    name: 'Mix And Match Shatter/Budder 28g (4 X 7G)',
    rate: '46/5',
    id: '1'
  },
  {
    isSoldOut: 86,
    title: '',
    totalReviews: 135,
    discount: '$200.00',
    price: '$102.00',
    unit: 'unit 1',
    volume: [],
    image: 'https://tinyurl.com/2s34y377',
    sale: 'Indica 70%',
    description:
      'Jungle Diamonds is a slightly indica dominant hybrid strain (60% indica/40% sativa) created through crossing the infamous Slurricane X Gorilla Glue #4 strains.',
    category: 'CONCENTRATES',
    name: 'Mix And Match Shatter/Budder 28g (4 X 7G)',
    rate: '46/5',
    id: '1'
  },
  {
    isSoldOut: 86,
    title: '',
    totalReviews: 135,
    discount: '$200.00',
    price: '$102.00',
    unit: 'unit 1',
    volume: [],
    image: 'https://tinyurl.com/2s34y377',
    sale: 'Indica 70%',
    description:
      'Jungle Diamonds is a slightly indica dominant hybrid strain (60% indica/40% sativa) created through crossing the infamous Slurricane X Gorilla Glue #4 strains.',
    category: 'CONCENTRATES',
    name: 'Mix And Match Shatter/Budder 28g (4 X 7G)',
    rate: '46/5',
    id: '1'
  },
  {
    isSoldOut: 86,
    title: '',
    totalReviews: 135,
    discount: '$200.00',
    price: '$102.00',
    unit: 'unit 1',
    volume: [],
    image: 'https://tinyurl.com/2s34y377',
    sale: 'Indica 70%',
    description:
      'Jungle Diamonds is a slightly indica dominant hybrid strain (60% indica/40% sativa) created through crossing the infamous Slurricane X Gorilla Glue #4 strains.',
    category: 'CONCENTRATES',
    name: 'Mix And Match Shatter/Budder 28g (4 X 7G)',
    rate: '46/5',
    id: '1'
  }
];

export const LIST_PRODUCT_CATEGORY_FILTER = [
  {
    name: 'Sales',
    value: 12
  },
  {
    name: 'Cannabis',
    value: 430
  },
  {
    name: 'Pre-Rolls',
    value: 40
  },
  {
    name: 'CBD Oil',
    value: 12
  },
  {
    name: 'Magic Mushrooms',
    value: 34
  },
  {
    name: 'Extracts',
    value: 26
  },
  {
    name: 'Edibles',
    value: 32
  },
  {
    name: 'Vape Pens',
    value: 12
  },
  {
    name: 'Accessories',
    value: 10
  },
  {
    name: 'Bath & Body',
    value: 8
  },
  {
    name: 'Bundles',
    value: 24
  },
  {
    name: 'Wholesale',
    value: 28
  }
];
