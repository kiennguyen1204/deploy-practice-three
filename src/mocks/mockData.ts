import { Product } from 'interfaces/item';
import { Cart } from '../interfaces/cart';

export const DATA_ITEM = {
  image: 'test-image.jpg',
  subtitle: 'Test Subtitle',
  title: 'Test Title',
  description: 'Test Description',
  to: '/test-link'
};

export const MOCK_PROPS = {
  image: '/sample_image.png',
  title: 'Sample Title',
  description: 'Sample Description',
  to: '/sample-link',
  shop: 'Sample Shop',
  icon: '/sample_icon.png'
};

export const MOCK_ITEM: Product = {
  id: '1',
  name: 'Product A',
  title: 'Product A Title',
  description: 'Description for Product A',
  sale: '30%',
  image: 'image-url-1',
  reviews: 15,
  isSoldOut: false,
  category: 'Category A',
  price: 100,
  discount: 10,
  volume: ['100ml', '200ml'],
  totalReviews: 20,
  quantity: 50
};

export const MOCK_DATA_CUSTOMER = {
  userName: 'Vikki Starr',
  avatar: '/src/assets/images/reviews/VikkiStarr.png',
  comment: `Absolutely love TopShelfBC; affordable on any budget and such fast delivery, straight to my door! I recommend them to all my friends and family for their 420 needs.`,
  star: 4,
  date: 'January 15, 2023'
};

export const DATA_CART = {
  total: 305.0,
  shipping: 50,
  payTotal: 355,
  discount: 0,
  cart: [
    {
      name: 'Mix And Match Shatter/Budder 28g (4 X 7G)',
      image: '/src/assets/images/products/snack_yellow.webp',
      price: 120.0,
      amount: 2,
      total: 240.0,
      integra: [
        {
          namePack: 'Add Integra Pack - 4g',
          amount: 1,
          price: 2.0,
          total: 2.0
        },
        {
          namePack: 'Add Integra Pack - 8g',
          amount: 1,
          price: 3.0,
          total: 3.0
        }
      ],
      subTotal: 245.0
    },
    {
      name: 'Mix And Match Shatter/Budder 28g (4 X 7G)',
      image: '/src/assets/images/products/snack_purple.webp',
      price: 13.0,
      amount: 5,
      total: 65.0,
      integra: [],
      subTotal: 65.0
    }
  ]
};

export const MOCK_PRODUCT: Product = {
  id: '1',
  name: 'Sample Product',
  title: 'Sample Title',
  description: 'This is a sample description.',
  sale: 'Sample Sale',
  image: 'sample_image.jpg',
  reviews: 10,
  isSoldOut: false,
  category: 'Sample Category',
  price: 100,
  discount: 20,
  volume: ['100g', '200g'],
  rate: '5',
  totalReviews: 50,
  quantity: 20
};

export const MOCK_CART: Cart[] = [
  {
    ...MOCK_PRODUCT,
    amount: 2,
    total: 200,
    subTotal: 180,
    quantity: 10,
    initialPrice: 90
  }
];
