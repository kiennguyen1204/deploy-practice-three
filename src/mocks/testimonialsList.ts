export const TESTIMONIALS_LIST = [
  {
    userName: 'Vikki Starr',
    avatar: '/src/assets/images/reviews/VikkiStarr.png',
    comment: `Absolutely love TopShelfBC; affordable on any budget and such fast delivery, straight to my door! I recommend them to all my friends and family for their 420 needs.`,
    star: 4,
    date: 'January 15, 2023'
  },
  {
    userName: 'Vikki Starr',
    avatar: '/src/assets/images/reviews/VikkiStarr.png',
    comment: `Absolutely love TopShelfBC; affordable on any budget and such fast delivery, straight to my door! I recommend them to all my friends and family for their 420 needs.`,
    star: 4,
    date: 'January 15, 2023'
  },
  {
    userName: 'Vikki Starr',
    avatar: '/src/assets/images/reviews/VikkiStarr.png',
    comment: `Absolutely love TopShelfBC; affordable on any budget and such fast delivery, straight to my door! I recommend them to all my friends and family for their 420 needs.`,
    star: 5,
    date: 'January 15, 2023'
  },
  {
    userName: 'Vikki Starr',
    avatar: '/src/assets/images/reviews/VikkiStarr.png',
    comment: `Absolutely love TopShelfBC; affordable on any budget and such fast delivery, straight to my door! I recommend them to all my friends and family for their 420 needs.`,
    star: 5,
    date: 'January 15, 2023'
  },
  {
    userName: 'Vikki Starr',
    avatar: '/src/assets/images/reviews/VikkiStarr.png',
    comment: `Absolutely love TopShelfBC; affordable on any budget and such fast delivery, straight to my door! I recommend them to all my friends and family for their 420 needs.`,
    star: 5,
    date: 'January 15, 2023'
  },
  {
    userName: 'Vikki Starr',
    avatar: '/src/assets/images/reviews/VikkiStarr.png',
    comment: `Absolutely love TopShelfBC; affordable on any budget and such fast delivery, straight to my door! I recommend them to all my friends and family for their 420 needs.`,
    star: 5,
    date: 'January 15, 2023'
  }
];
