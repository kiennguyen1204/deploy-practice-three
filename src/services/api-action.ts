import axios from 'axios';

// constants
import { PRODUCTS_API } from 'constants/api';
import { Product } from 'interfaces/item';

/**
 * Retrieves a product by its ID from the API.
 * @param {string} id - The ID of the product.
 * @returns {Promise<Product | null>} - The product with the specified ID, or null if not found.
 * @throws {Error} - If there's an error retrieving the product.
 */
export const getProductById = async (id: string): Promise<Product | null> => {
  try {
    const url = `${PRODUCTS_API}/${id}`;

    const response = await axios.get(url, {
      headers: { 'content-type': 'application/json' }
    });

    return response.data;
  } catch (error) {
    throw error;
  }
};

export const getProducts = async (): Promise<Product[]> => {
  try {
    const response = await axios.get(PRODUCTS_API, {
      headers: { 'content-type': 'application/json' }
    });

    return response.data;
  } catch (error) {
    throw error;
  }
};
