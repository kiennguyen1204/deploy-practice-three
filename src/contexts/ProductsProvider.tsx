import { ReactNode, createContext, useCallback, useEffect, useMemo, useState } from 'react';

// interface
import { Product } from 'interfaces/item';

// services
import { addToCart, getProducts } from 'services/api-action';

interface ProductProviderProps {
  children: ReactNode;
}

export interface ProductContextProps {
  products: Product[];
  onGetProducts: () => void;
  onAddToCart: (product: Product) => void;
}

export const ProductContext = createContext<ProductContextProps>({
  products: [],
  onGetProducts: () => {},
  onAddToCart: () => undefined
});

const ProductsProvider: React.FC<ProductProviderProps> = ({ children }) => {
  const [products, setProducts] = useState<Product[]>([]);

  const fetchListProducts = async () => {
    const productsData = await getProducts();
    setProducts(productsData);
  };

  useEffect(() => {
    fetchListProducts();
  }, []);

  const handleAddToCart = useCallback(async (product: Product): Promise<void> => {
    const data = await addToCart(product);
    setProducts((prevCart: Product[]) => [...prevCart, data]);
  }, []);

  const cartContextValue = useMemo(
    () => ({
      products,
      onGetProducts: fetchListProducts,
      onAddToCart: handleAddToCart
    }),
    [products]
  );

  return <ProductContext.Provider value={cartContextValue}>{children}</ProductContext.Provider>;
};

export default ProductsProvider;
