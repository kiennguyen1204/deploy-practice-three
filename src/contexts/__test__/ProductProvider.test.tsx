import { useContext } from 'react';

import { act, fireEvent, render, screen } from '@testing-library/react';
import ProductsProvider, { ProductContext } from '../ProductsProvider';

// services
import * as apiAction from 'services/api-action';

// mocks
import { MOCK_PRODUCT } from 'mocks/mockData';

jest.mock('services/api-action', () => ({
  __esModule: true,
  getProducts: jest.fn(),
  addToCart: jest.fn()
}));

describe('ProductsProvider', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render children', () => {
    render(
      <ProductsProvider>
        <div data-testid='child'>Child Component</div>
      </ProductsProvider>
    );

    expect(screen.getByTestId('child')).toBeInTheDocument();
  });

  it('should fetch products on mount', async () => {
    // Mock the getProducts function
    jest.spyOn(apiAction, 'getProducts').mockResolvedValue([MOCK_PRODUCT]);

    render(
      <ProductsProvider>
        <div data-testid='child'>Child Component</div>
      </ProductsProvider>
    );

    // Wait for the useEffect to fetch the products
    await screen.findByText('Child Component');

    // Assert that the products are set correctly
    expect(screen.getByTestId('child')).toBeInTheDocument();
  });

  it('should add a product to the cart', async () => {
    const postProductMock = jest.spyOn(apiAction, 'addToCart');
    postProductMock.mockResolvedValue(MOCK_PRODUCT);

    const MockChildren = () => {
      const { onAddToCart } = useContext(ProductContext);

      return <button title='Add to cart' onClick={() => onAddToCart(MOCK_PRODUCT)} />;
    };

    const { getByRole } = render(
      <ProductsProvider>
        <MockChildren />
      </ProductsProvider>
    );

    const button = getByRole('button');

    await act(() => {
      fireEvent.click(button);
    });
    expect(postProductMock).toHaveBeenCalledWith(MOCK_PRODUCT);
  });
});
