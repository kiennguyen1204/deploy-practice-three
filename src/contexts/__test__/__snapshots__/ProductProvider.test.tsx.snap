// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsProvider should render the component and match the snapshot 1`] = `
<div>
  <div>
    Test
  </div>
</div>
`;
